# Capstone Project - Data cleaning

In [1]:
import pandas as pd
import numpy as np
import pickle
import matplotlib.pyplot as plt
pd.set_option('display.max_columns',50)

### Reading in scraped shoe data

In [2]:
df = pd.read_csv("shoe_data.csv")

In [3]:
print(f'No. of shoes: {df.shape[0]}')
print(f'No. of features: {df.shape[1]}')

No. of shoes: 2066
No. of features: 35


In [4]:
df

name         brand  core_score  \
0                Nike Air Zoom Pegasus 35          Nike        94.0   
1                         Brooks Ghost 11        Brooks        94.0   
2                     Asics Gel Kayano 25         Asics        95.0   
3                Brooks Adrenaline GTS 19        Brooks        94.0   
4                      Adidas Ultra Boost        Adidas        90.0   
5                                On Cloud            On        90.0   
6                     Asics Gel Venture 6         Asics        94.0   
7                    Hoka One One Bondi 6  Hoka One One        91.0   
8                      Nike Downshifter 7          Nike        94.0   
9                      Asics Gel Excite 6         Asics        95.0   
10                  Merrell Trail Glove 4       Merrell        93.0   
11               Nike Air Zoom Pegasus 36          Nike        91.0   
12                 Nike Air Zoom Winflo 6          Nike        90.0   
13               Brooks Adrenaline GTS 18        Brooks        94.0   
14                        Xero Shoes Prio    Xero Shoes        99.0   
15                     New Balance 990 v4   New Balance        92.0   
16                Nike Zoom Pegasus Turbo          Nike        92.0   
17                    Asics Gel Contend 5         Asics        92.0   
18                    Asics Gel Nimbus 21         Asics        91.0   
19                Nike Epic React Flyknit          Nike        89.0   
20                          Nike Zoom Fly          Nike        93.0   
21                      Nike Revolution 4          Nike        94.0   
22                     Brooks Glycerin 16        Brooks        93.0   
23                     Brooks Glycerin 17        Brooks        94.0   
24                 Hoka One One Clifton 5  Hoka One One        93.0   
25              Nike Free RN Flyknit 2018          Nike        94.0   
26              Nike Epic React Flyknit 2          Nike        89.0   
27                   Adidas Pure Boost Go        Adidas        94.0   
28                        Altra Escalante         Altra        95.0   
29                   Adidas Ultraboost 19        Adidas        95.0   
...                                   ...           ...         ...   
2036              Dynafit Feline Vertical           NaN         NaN   
2037             Salomon Speedcross 5 GTX           NaN         NaN   
2038                   Mizuno WaveKnit C1           NaN         NaN   
2039                   Scott Kinabalu GTX           NaN         NaN   
2040                    Dynafit Speed MTN           NaN         NaN   
2041              Mizuno Wave Mujin 5 GTX           NaN         NaN   
2042                Mizuno Wave Ibuki GTX           NaN         NaN   
2043             Under Armour Horizon BPF           NaN         NaN   
2044                       Puma Speed 500           NaN         NaN   
2045                 Mizuno Wave Daichi 4           NaN         NaN   
2046                       Scott Kinabalu           NaN         NaN   
2047                   Mizuno WaveKnit S1           NaN         NaN   
2048                     Skechers GOrun 7           NaN         NaN   
2049             Mizuno Wave Daichi 4 GTX           NaN         NaN   
2050                  Mizuno Wave Polaris           NaN         NaN   
2051                 Mizuno Wave Equate 3           NaN         NaN   
2052             Dynafit Trailbreaker GTX           NaN         NaN   
2053                  Mizuno Wave Prodigy           NaN         NaN   
2054          Dynafit Feline Vertical Pro           NaN         NaN   
2055          Saucony Zealot ISO 3 Reflex           NaN         NaN   
2056  New Balance FuelCore Sonic Viz Pack           NaN         NaN   
2057    New Balance Fresh Foam Cruz Sport           NaN         NaN   
2058                  Icebug DTS 3 BUGrip           NaN         NaN   
2059                Icebug DTS 3 RB9X GTX           NaN         NaN   
2060                Dynafit Speed MTN GTX           NaN      

### Data cleaning

#### Remove duplicates

In [5]:
len(set(df.name))

2058

In [6]:
df.duplicated().value_counts()

False    2058
True        8
dtype: int64

In [7]:
df.drop_duplicates(inplace=True) 

In [8]:
len(df.name)

2058

#### Fill up null values

In [9]:
df.isnull().sum()

name                      0
brand                   114
core_score              114
user_review_count       114
expert_review_count     865
user_rating             114
expert_rating           865
user_5star              117
user_4star              117
user_3star              117
user_2star              117
user_1star              117
discontinued              0
terrain                   0
arch_support              5
use                     116
waterproof             1948
water_repellent        2004
maximalist             1984
minimalist             2025
triathlon              2042
price                     3
weight                  112
toe_drop                147
heel_height             669
forefoot_height         670
width                     0
good_summary            114
bad_summary             114
summary                 115
info                     92
size_fit                  2
outsole                   1
midsole                   2
upper                     5
dtype: int64

- 865 shoes (42.1%) were not reviewed/ rated by Experts. <br>
- 114 shoes (5.7%) were new at the time of scraping and had not been reviewed yet. <br>

In [10]:
df.dropna().shape

(0, 35)

All shoes have at least 1 missing feature.

In [11]:
# Fill up NaN values accordingly.
df['waterproof'] = df['waterproof'].fillna('Not waterproof')
df['water_repellent'] = df['water_repellent'].fillna('Not water repellent')
df['maximalist'] = df['maximalist'].fillna('Not maximalist')
df['minimalist'] = df['minimalist'].fillna('Not minimalist')
df['triathlon'] = df['triathlon'].fillna('Not triathlon')

In [12]:
# Check
df.maximalist.value_counts()

Not maximalist    1984
Maximalist          74
Name: maximalist, dtype: int64

In [13]:
df.isnull().sum()

name                     0
brand                  114
core_score             114
user_review_count      114
expert_review_count    865
user_rating            114
expert_rating          865
user_5star             117
user_4star             117
user_3star             117
user_2star             117
user_1star             117
discontinued             0
terrain                  0
arch_support             5
use                    116
waterproof               0
water_repellent          0
maximalist               0
minimalist               0
triathlon                0
price                    3
weight                 112
toe_drop               147
heel_height            669
forefoot_height        670
width                    0
good_summary           114
bad_summary            114
summary                115
info                    92
size_fit                 2
outsole                  1
midsole                  2
upper                    5
dtype: int64

#### Drop rows with null values

In [14]:
df = df[pd.notnull(df['heel_height'])]

In [15]:
df.isnull().sum()

name                     0
brand                   63
core_score              63
user_review_count       63
expert_review_count    445
user_rating             63
expert_rating          445
user_5star              63
user_4star              63
user_3star              63
user_2star              63
user_1star              63
discontinued             0
terrain                  0
arch_support             2
use                     16
waterproof               0
water_repellent          0
maximalist               0
minimalist               0
triathlon                0
price                    3
weight                  12
toe_drop                 1
heel_height              0
forefoot_height          1
width                    0
good_summary            63
bad_summary             63
summary                 63
info                    58
size_fit                 0
outsole                  1
midsole                  2
upper                    4
dtype: int64

In [16]:
df = df[pd.notnull(df['forefoot_height'])]

In [17]:
df.isnull().sum()

name                     0
brand                   62
core_score              62
user_review_count       62
expert_review_count    444
user_rating             62
expert_rating          444
user_5star              62
user_4star              62
user_3star              62
user_2star              62
user_1star              62
discontinued             0
terrain                  0
arch_support             2
use                     16
waterproof               0
water_repellent          0
maximalist               0
minimalist               0
triathlon                0
price                    3
weight                  12
toe_drop                 0
heel_height              0
forefoot_height          0
width                    0
good_summary            62
bad_summary             62
summary                 62
info                    58
size_fit                 0
outsole                  1
midsole                  2
upper                    4
dtype: int64

In [18]:
df.shape

(1388, 35)

In [19]:
# Drop the new shoes which have not been reviewed.
df = df[pd.notnull(df['user_review_count'])]

In [20]:
df.isnull().sum()

name                     0
brand                    0
core_score               0
user_review_count        0
expert_review_count    382
user_rating              0
expert_rating          382
user_5star               0
user_4star               0
user_3star               0
user_2star               0
user_1star               0
discontinued             0
terrain                  0
arch_support             2
use                     15
waterproof               0
water_repellent          0
maximalist               0
minimalist               0
triathlon                0
price                    0
weight                  12
toe_drop                 0
heel_height              0
forefoot_height          0
width                    0
good_summary             0
bad_summary              0
summary                  0
info                    49
size_fit                 0
outsole                  1
midsole                  2
upper                    4
dtype: int64

In [21]:
df.shape

(1326, 35)

In [22]:
# Drop shoes which experts did not give any reviews for
df = df[pd.notnull(df['expert_review_count'])]

In [23]:
df.isnull().sum()

name                    0
brand                   0
core_score              0
user_review_count       0
expert_review_count     0
user_rating             0
expert_rating           0
user_5star              0
user_4star              0
user_3star              0
user_2star              0
user_1star              0
discontinued            0
terrain                 0
arch_support            0
use                     9
waterproof              0
water_repellent         0
maximalist              0
minimalist              0
triathlon               0
price                   0
weight                  6
toe_drop                0
heel_height             0
forefoot_height         0
width                   0
good_summary            0
bad_summary             0
summary                 0
info                   35
size_fit                0
outsole                 0
midsole                 2
upper                   3
dtype: int64

In [24]:
df.shape

(944, 35)

In [25]:
# Drop rows with null values for use
df = df[pd.notnull(df['use'])]

In [26]:
df.isnull().sum()

name                    0
brand                   0
core_score              0
user_review_count       0
expert_review_count     0
user_rating             0
expert_rating           0
user_5star              0
user_4star              0
user_3star              0
user_2star              0
user_1star              0
discontinued            0
terrain                 0
arch_support            0
use                     0
waterproof              0
water_repellent         0
maximalist              0
minimalist              0
triathlon               0
price                   0
weight                  0
toe_drop                0
heel_height             0
forefoot_height         0
width                   0
good_summary            0
bad_summary             0
summary                 0
info                   33
size_fit                0
outsole                 0
midsole                 2
upper                   3
dtype: int64

In [27]:
df.shape

(935, 35)

In [28]:
# Drop rows with null values for info
df = df[pd.notnull(df['info'])]

In [29]:
df.isnull().sum()

name                   0
brand                  0
core_score             0
user_review_count      0
expert_review_count    0
user_rating            0
expert_rating          0
user_5star             0
user_4star             0
user_3star             0
user_2star             0
user_1star             0
discontinued           0
terrain                0
arch_support           0
use                    0
waterproof             0
water_repellent        0
maximalist             0
minimalist             0
triathlon              0
price                  0
weight                 0
toe_drop               0
heel_height            0
forefoot_height        0
width                  0
good_summary           0
bad_summary            0
summary                0
info                   0
size_fit               0
outsole                0
midsole                2
upper                  3
dtype: int64

In [30]:
df.shape

(902, 35)

In [31]:
# Drop rows with null values for midsole
df = df[pd.notnull(df['midsole'])]

In [32]:
df.isnull().sum()

name                   0
brand                  0
core_score             0
user_review_count      0
expert_review_count    0
user_rating            0
expert_rating          0
user_5star             0
user_4star             0
user_3star             0
user_2star             0
user_1star             0
discontinued           0
terrain                0
arch_support           0
use                    0
waterproof             0
water_repellent        0
maximalist             0
minimalist             0
triathlon              0
price                  0
weight                 0
toe_drop               0
heel_height            0
forefoot_height        0
width                  0
good_summary           0
bad_summary            0
summary                0
info                   0
size_fit               0
outsole                0
midsole                0
upper                  2
dtype: int64

In [33]:
df.shape

(900, 35)

In [34]:
# Drop rows with null values for upper
df = df[pd.notnull(df['upper'])]

In [37]:
df.isnull().sum()

name                   0
brand                  0
core_score             0
user_review_count      0
expert_review_count    0
user_rating            0
expert_rating          0
user_5star             0
user_4star             0
user_3star             0
user_2star             0
user_1star             0
discontinued           0
terrain                0
arch_support           0
use                    0
waterproof             0
water_repellent        0
maximalist             0
minimalist             0
triathlon              0
price                  0
weight                 0
toe_drop               0
heel_height            0
forefoot_height        0
width                  0
good_summary           0
bad_summary            0
summary                0
info                   0
size_fit               0
outsole                0
midsole                0
upper                  0
dtype: int64

In [35]:
df = df.reset_index(drop=True)

In [36]:
df

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike        94.0   
1                   Brooks Ghost 11          Brooks        94.0   
2               Asics Gel Kayano 25           Asics        95.0   
3               Asics Gel Venture 6           Asics        94.0   
4              Hoka One One Bondi 6    Hoka One One        91.0   
5             Merrell Trail Glove 4         Merrell        93.0   
6          Nike Air Zoom Pegasus 36            Nike        91.0   
7          Brooks Adrenaline GTS 18          Brooks        94.0   
8           Nike Zoom Pegasus Turbo            Nike        92.0   
9               Asics Gel Nimbus 21           Asics        91.0   
10               Brooks Glycerin 16          Brooks        93.0   
11               Brooks Glycerin 17          Brooks        94.0   
12           Hoka One One Clifton 5    Hoka One One        93.0   
13        Nike Free RN Flyknit 2018            Nike        94.0   
14        Nike Epic React Flyknit 2            Nike        89.0   
15             Adidas Pure Boost Go          Adidas        94.0   
16             Adidas Ultraboost 19          Adidas        95.0   
17               Adidas Solar Boost          Adidas        94.0   
18            Nike Zoom Fly Flyknit            Nike        92.0   
19              Altra Escalante 1.5           Altra        94.0   
20         Hoka One One Speedgoat 3    Hoka One One        97.0   
21         Nike Air Zoom Pegasus 34            Nike        92.0   
22                Hoka One One Cavu    Hoka One One        92.0   
23                Nike Free RN 2018            Nike        92.0   
24             Mizuno Wave Rider 22          Mizuno        94.0   
25                     Brooks Revel          Brooks        95.0   
26   New Balance Fresh Foam 1080 v9     New Balance        95.0   
27                  Brooks Ghost 10          Brooks        95.0   
28               Saucony Kinvara 10         Saucony        93.0   
29             Asics Gel Quantum 90           Asics        92.0   
..                              ...             ...         ...   
868       Reebok All Terrain Thrill          Reebok        72.0   
869              Altra Paradigm 1.5           Altra        83.0   
870              Altra Instinct 3.5           Altra        83.0   
871        Brooks Adrenaline ASR 12          Brooks        83.0   
872         Asics Gel FujiTrainer 3           Asics        83.0   
873     Asics Gel FujiTrabuco 5 GTX           Asics        83.0   
874  Reebok All Terrain Thunder 2.0          Reebok        82.0   
875              Inov-8 Mudclaw 265          Inov-8        82.0   
876                   Zoot Ovwa 2.0            Zoot        82.0   
877      Salomon S-Lab Spikecross 3         Salomon        82.0   
878        Under Armour Charge RC 2    Under Armour        82.0   
879        Adidas Adizero Feather 4          Adidas        70.0   
880    Under Armour Micro G Nitrous    Under Armour        81.0   
881       Merrell All Out Terra Ice         Merrell        81.0   
882             Salomon S-Lab Sonic         Salomon        81.0   
883             Asics FuzeX Rush CM           Asics        81.0   
884              Puma Faas 100 R v3            Puma        80.0   
885         Asics Gel Kenun Shinkai           Asics        80.0   
886     The North Face Ultra TR III  The North Face        80.0   
887                 Inov-8 Oroc 340          Inov-8        79.0   
888              Zoot Advantage 3.0            Zoot        79.0   
889         Inov-8 Road-X-Treme 220          Inov-8        79.0   
890         Adidas Pure Boost Chill          Adidas        79.0   
891                      Zoot Diego            Zoot        78.0   
892              Saucony Ride 9 GTX         Saucony        78.0   
893              Skechers GOtrail 2        Skechers        77.0   
894             Adidas Adizero XT 5          Adidas        76.0   
895           Mizuno Wave Cruise 10          Mizuno        64.0   
896        

In [38]:
# Create DataFrames with split value columns.
split_weight = df["weight"].str.split(";", n=1, expand=True)
split_toe_drop = df["toe_drop"].str.split(";", n=1, expand=True)
split_heel_height = df["heel_height"].str.split(";", n=1, expand=True)
split_forefoot_height = df["forefoot_height"].str.split(";", n=1, expand=True)
split_width = df["width"].str.split(";", n=1, expand=True)

In [39]:
split_width.head(3)

0                               1
0  Men: Narrow, Standard, Wide, Extra wide          Women: Standard, Wide;
1  Men: Narrow, Standard, Wide, Extra wide  Women: Narrow, Standard, Wide;
2  Men: Narrow, Standard, Wide, Extra wide  Women: Narrow, Standard, Wide;

In [40]:
split_weight[0].value_counts()

Men: 255g    24
Men: 275g    23
Men: 298g    23
Men: 295g    20
Men: 292g    20
Men: 269g    18
Men: 309g    17
Men: 283g    15
Men: 312g    15
Men: 306g    14
Men: 318g    14
Men: 238g    14
Men: 266g    14
Men: 300g    13
Men: 281g    13
Men: 244g    13
Men: 320g    12
Men: 323g    12
Men: 272g    12
Men: 315g    12
Men: 258g    12
Men: 227g    12
Men: 303g    12
Men: 289g    11
Men: 232g    11
Men: 264g    11
Men: 230g    11
Men: 224g    11
Men: 261g    11
Men: 278g    10
             ..
Men: 202g     1
Men: 130g     1
Men: 79g      1
Men: 368g     1
Men: 328g     1
Men: 361g     1
Men: 510g     1
Men: 277g     1
Men: 391g     1
Men: 115g     1
Men: 400g     1
Men: 370g     1
Men: 102g     1
Men: 410g     1
Men: 116g     1
Men: 136g     1
Men: 128g     1
Men: 383g     1
Men: 291g     1
Men: 217g     1
Men: 145g     1
Men: 195g     1
Men: 268g     1
Men: 214g     1
Men: 416g     1
Men: 263g     1
Men: 248g     1
Men: 302g     1
Men: 307g     1
Men: 172g     1
Name: 0, Length: 185, dt

In [41]:
split_weight[1].value_counts()

Women: 255g;    42
Women: 235g;    27
                25
Women: 190g;    21
Women: 227g;    18
Women: 230g;    17
Women: 252g;    17
Women: 244g;    17
Women: 241g;    17
Women: 261g;    16
Women: 215g;    16
Women: 198g;    15
Women: 232g;    15
Women: 187g;    15
Women: 249g;    15
Women: 213g;    15
Women: 247g;    14
Women: 218g;    14
Women: 275g;    14
Women: 238g;    14
Women: 258g;    13
Women: 184g;    12
Women: 210g;    12
Women: 207g;    12
Women: 204g;    11
Women: 289g;    11
Women: 269g;    11
Women: 266g;    11
Women: 281g;    10
Women: 221g;    10
                ..
Women: 328g;     1
Women: 138g;     1
Women: 311g;     1
Women: 279g;     1
Women: 203g;     1
Women: 357g;     1
Women: 375g;     1
Women: 239g;     1
Women: 286g;     1
Women: 324g;     1
Women: 350g;     1
Women: 273g;     1
Women: 80g;      1
Women: 268g;     1
Women: 209g;     1
Women: 111g;     1
Women: 168g;     1
Women: 510g;     1
Women: 199g;     1
Women: 354g;     1
Women: 115g;     1
Women: 116g;

In [56]:
women_is_empty_str_mask = split_weight[1] == ""
len(split_weight[women_is_empty_str_mask])

25

In [43]:
split_weight[0].str.contains('Women').any().sum()

1

In [44]:
df['weight_m'] = split_weight[0].map(lambda x: x.replace("Men: ","").replace("g","") if "Men" in x else 0)
df['weight_w'] = split_weight[1].map(lambda x: x.replace("Women: ","").replace("g;","") if "Women" in x else 0)

In [45]:
women_is_empty_str_mask2 = df['weight_w'] == ""
len(df[women_is_empty_str_mask2])

0

In [46]:
df['weight_m'].str.contains('Women').any()

False

In [47]:
mask = df['weight_w'] == 0
len(df[mask])

25

In [48]:
df.shape

(898, 37)

In [49]:
df = df[df['weight_w'] != 0]
df = df.reset_index(drop=True)
df

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike        94.0   
1                   Brooks Ghost 11          Brooks        94.0   
2               Asics Gel Kayano 25           Asics        95.0   
3               Asics Gel Venture 6           Asics        94.0   
4              Hoka One One Bondi 6    Hoka One One        91.0   
5             Merrell Trail Glove 4         Merrell        93.0   
6          Nike Air Zoom Pegasus 36            Nike        91.0   
7          Brooks Adrenaline GTS 18          Brooks        94.0   
8           Nike Zoom Pegasus Turbo            Nike        92.0   
9               Asics Gel Nimbus 21           Asics        91.0   
10               Brooks Glycerin 16          Brooks        93.0   
11               Brooks Glycerin 17          Brooks        94.0   
12           Hoka One One Clifton 5    Hoka One One        93.0   
13        Nike Free RN Flyknit 2018            Nike        94.0   
14        Nike Epic React Flyknit 2            Nike        89.0   
15             Adidas Pure Boost Go          Adidas        94.0   
16             Adidas Ultraboost 19          Adidas        95.0   
17               Adidas Solar Boost          Adidas        94.0   
18            Nike Zoom Fly Flyknit            Nike        92.0   
19              Altra Escalante 1.5           Altra        94.0   
20         Hoka One One Speedgoat 3    Hoka One One        97.0   
21         Nike Air Zoom Pegasus 34            Nike        92.0   
22                Hoka One One Cavu    Hoka One One        92.0   
23                Nike Free RN 2018            Nike        92.0   
24             Mizuno Wave Rider 22          Mizuno        94.0   
25                     Brooks Revel          Brooks        95.0   
26   New Balance Fresh Foam 1080 v9     New Balance        95.0   
27                  Brooks Ghost 10          Brooks        95.0   
28               Saucony Kinvara 10         Saucony        93.0   
29             Asics Gel Quantum 90           Asics        92.0   
..                              ...             ...         ...   
843       Reebok All Terrain Thrill          Reebok        72.0   
844              Altra Paradigm 1.5           Altra        83.0   
845              Altra Instinct 3.5           Altra        83.0   
846        Brooks Adrenaline ASR 12          Brooks        83.0   
847         Asics Gel FujiTrainer 3           Asics        83.0   
848     Asics Gel FujiTrabuco 5 GTX           Asics        83.0   
849  Reebok All Terrain Thunder 2.0          Reebok        82.0   
850              Inov-8 Mudclaw 265          Inov-8        82.0   
851                   Zoot Ovwa 2.0            Zoot        82.0   
852      Salomon S-Lab Spikecross 3         Salomon        82.0   
853        Under Armour Charge RC 2    Under Armour        82.0   
854        Adidas Adizero Feather 4          Adidas        70.0   
855    Under Armour Micro G Nitrous    Under Armour        81.0   
856       Merrell All Out Terra Ice         Merrell        81.0   
857             Salomon S-Lab Sonic         Salomon        81.0   
858             Asics FuzeX Rush CM           Asics        81.0   
859              Puma Faas 100 R v3            Puma        80.0   
860         Asics Gel Kenun Shinkai           Asics        80.0   
861     The North Face Ultra TR III  The North Face        80.0   
862                 Inov-8 Oroc 340          Inov-8        79.0   
863              Zoot Advantage 3.0            Zoot        79.0   
864         Inov-8 Road-X-Treme 220          Inov-8        79.0   
865         Adidas Pure Boost Chill          Adidas        79.0   
866                      Zoot Diego            Zoot        78.0   
867              Saucony Ride 9 GTX         Saucony        78.0   
868              Skechers GOtrail 2        Skechers        77.0   
869             Adidas Adizero XT 5          Adidas        76.0   
870           Mizuno Wave Cruise 10          Mizuno        64.0   
871        

In [50]:
split_toe_drop[0].str.contains('Women').any().sum()

1

In [57]:
mask = split_toe_drop[1] == ""
len(split_toe_drop[mask])

23

In [52]:
split_heel_height[0].str.contains('Women').any().sum()

1

In [58]:
mask = split_heel_height[1] == ""
len(split_heel_height[mask])

25

In [54]:
split_forefoot_height[0].str.contains('Women').any().sum()

1

In [59]:
mask = split_forefoot_height[1] == ""
len(split_forefoot_height[mask])

25

In [60]:
split_width[0].str.contains('Women').any().sum()

0

In [61]:
mask = split_width[1] == ""
len(split_width[mask])

0

In [62]:
# Split shoe features into men's and women's.

df['toe_drop_m'] = split_toe_drop[0].map(lambda x: x.replace("Men: ","").replace("mm","") if "Men" in x else 0)
df['toe_drop_w'] = split_toe_drop[1].map(lambda x: x.replace("Women: ","").replace("mm;","") if "Women" in x else 0)

df['heel_height_m'] = split_heel_height[0].map(lambda x: x.replace("Men: ","").replace("mm","") if "Men" in x else 0)
df['heel_height_w'] = split_heel_height[1].map(lambda x: x.replace("Women: ","").replace("mm;","") if "Women" in x else 0)

df['forefoot_height_m'] = split_forefoot_height[0].map(lambda x: x.replace("Men: ","").replace("mm","") if "Men" in x else 0)
df['forefoot_height_w'] = split_forefoot_height[1].map(lambda x: x.replace("Women: ","").replace("mm;","") if "Women" in x else 0)

df['width_m'] = split_width[0].map(lambda x: x.replace("Men: ","") if "Men" in x else 0)
df['width_w'] = split_width[1].map(lambda x: x.replace("Women: ","").replace(";","") if "Women" in x else 0)

In [63]:
df

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike        94.0   
1                   Brooks Ghost 11          Brooks        94.0   
2               Asics Gel Kayano 25           Asics        95.0   
3               Asics Gel Venture 6           Asics        94.0   
4              Hoka One One Bondi 6    Hoka One One        91.0   
5             Merrell Trail Glove 4         Merrell        93.0   
6          Nike Air Zoom Pegasus 36            Nike        91.0   
7          Brooks Adrenaline GTS 18          Brooks        94.0   
8           Nike Zoom Pegasus Turbo            Nike        92.0   
9               Asics Gel Nimbus 21           Asics        91.0   
10               Brooks Glycerin 16          Brooks        93.0   
11               Brooks Glycerin 17          Brooks        94.0   
12           Hoka One One Clifton 5    Hoka One One        93.0   
13        Nike Free RN Flyknit 2018            Nike        94.0   
14        Nike Epic React Flyknit 2            Nike        89.0   
15             Adidas Pure Boost Go          Adidas        94.0   
16             Adidas Ultraboost 19          Adidas        95.0   
17               Adidas Solar Boost          Adidas        94.0   
18            Nike Zoom Fly Flyknit            Nike        92.0   
19              Altra Escalante 1.5           Altra        94.0   
20         Hoka One One Speedgoat 3    Hoka One One        97.0   
21         Nike Air Zoom Pegasus 34            Nike        92.0   
22                Hoka One One Cavu    Hoka One One        92.0   
23                Nike Free RN 2018            Nike        92.0   
24             Mizuno Wave Rider 22          Mizuno        94.0   
25                     Brooks Revel          Brooks        95.0   
26   New Balance Fresh Foam 1080 v9     New Balance        95.0   
27                  Brooks Ghost 10          Brooks        95.0   
28               Saucony Kinvara 10         Saucony        93.0   
29             Asics Gel Quantum 90           Asics        92.0   
..                              ...             ...         ...   
843       Reebok All Terrain Thrill          Reebok        72.0   
844              Altra Paradigm 1.5           Altra        83.0   
845              Altra Instinct 3.5           Altra        83.0   
846        Brooks Adrenaline ASR 12          Brooks        83.0   
847         Asics Gel FujiTrainer 3           Asics        83.0   
848     Asics Gel FujiTrabuco 5 GTX           Asics        83.0   
849  Reebok All Terrain Thunder 2.0          Reebok        82.0   
850              Inov-8 Mudclaw 265          Inov-8        82.0   
851                   Zoot Ovwa 2.0            Zoot        82.0   
852      Salomon S-Lab Spikecross 3         Salomon        82.0   
853        Under Armour Charge RC 2    Under Armour        82.0   
854        Adidas Adizero Feather 4          Adidas        70.0   
855    Under Armour Micro G Nitrous    Under Armour        81.0   
856       Merrell All Out Terra Ice         Merrell        81.0   
857             Salomon S-Lab Sonic         Salomon        81.0   
858             Asics FuzeX Rush CM           Asics        81.0   
859              Puma Faas 100 R v3            Puma        80.0   
860         Asics Gel Kenun Shinkai           Asics        80.0   
861     The North Face Ultra TR III  The North Face        80.0   
862                 Inov-8 Oroc 340          Inov-8        79.0   
863              Zoot Advantage 3.0            Zoot        79.0   
864         Inov-8 Road-X-Treme 220          Inov-8        79.0   
865         Adidas Pure Boost Chill          Adidas        79.0   
866                      Zoot Diego            Zoot        78.0   
867              Saucony Ride 9 GTX         Saucony        78.0   
868              Skechers GOtrail 2        Skechers        77.0   
869             Adidas Adizero XT 5          Adidas        76.0   
870           Mizuno Wave Cruise 10          Mizuno        64.0   
871        

In [68]:
df = df[df['toe_drop_m'] != 0]
df = df.reset_index(drop=True)
df.shape

(872, 45)

In [71]:
df = df[df['toe_drop_w'] != 0]
df = df.reset_index(drop=True)
df.shape

(850, 45)

In [75]:
df = df[df['heel_height_m'] != 0]
df = df.reset_index(drop=True)
df.shape

(849, 45)

In [77]:
df = df[df['heel_height_w'] != 0]
df = df.reset_index(drop=True)
df.shape

(848, 45)

In [82]:
# Check...
print(len(df[df['weight_m'] == 0]))
print(len(df[df['weight_w'] == 0]))
print(len(df[df['toe_drop_m'] == 0]))
print(len(df[df['toe_drop_w'] == 0]))
print(len(df[df['heel_height_m'] == 0]))
print(len(df[df['heel_height_w'] == 0]))
print(len(df[df['forefoot_height_m'] == 0]))
print(len(df[df['forefoot_height_w'] == 0]))
print(len(df[df['width_m'] == 0]))
print(len(df[df['width_w'] == 0]))

0
0
0
0
0
0
0
0
0
0


In [83]:
# Clean up number of star rating percentages.
df["user_5star"] = df["user_5star"].str.replace("%","")
df["user_4star"] = df["user_4star"].str.replace("%","")
df["user_3star"] = df["user_3star"].str.replace("%","")
df["user_2star"] = df["user_2star"].str.replace("%","")
df["user_1star"] = df["user_1star"].str.replace("%","")

In [84]:
df

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike        94.0   
1                   Brooks Ghost 11          Brooks        94.0   
2               Asics Gel Kayano 25           Asics        95.0   
3               Asics Gel Venture 6           Asics        94.0   
4              Hoka One One Bondi 6    Hoka One One        91.0   
5             Merrell Trail Glove 4         Merrell        93.0   
6          Nike Air Zoom Pegasus 36            Nike        91.0   
7          Brooks Adrenaline GTS 18          Brooks        94.0   
8           Nike Zoom Pegasus Turbo            Nike        92.0   
9               Asics Gel Nimbus 21           Asics        91.0   
10               Brooks Glycerin 16          Brooks        93.0   
11               Brooks Glycerin 17          Brooks        94.0   
12           Hoka One One Clifton 5    Hoka One One        93.0   
13        Nike Free RN Flyknit 2018            Nike        94.0   
14        Nike Epic React Flyknit 2            Nike        89.0   
15             Adidas Pure Boost Go          Adidas        94.0   
16             Adidas Ultraboost 19          Adidas        95.0   
17               Adidas Solar Boost          Adidas        94.0   
18            Nike Zoom Fly Flyknit            Nike        92.0   
19              Altra Escalante 1.5           Altra        94.0   
20         Hoka One One Speedgoat 3    Hoka One One        97.0   
21         Nike Air Zoom Pegasus 34            Nike        92.0   
22                Hoka One One Cavu    Hoka One One        92.0   
23                Nike Free RN 2018            Nike        92.0   
24             Mizuno Wave Rider 22          Mizuno        94.0   
25                     Brooks Revel          Brooks        95.0   
26   New Balance Fresh Foam 1080 v9     New Balance        95.0   
27                  Brooks Ghost 10          Brooks        95.0   
28               Saucony Kinvara 10         Saucony        93.0   
29             Asics Gel Quantum 90           Asics        92.0   
..                              ...             ...         ...   
818       Reebok All Terrain Thrill          Reebok        72.0   
819              Altra Paradigm 1.5           Altra        83.0   
820              Altra Instinct 3.5           Altra        83.0   
821        Brooks Adrenaline ASR 12          Brooks        83.0   
822         Asics Gel FujiTrainer 3           Asics        83.0   
823     Asics Gel FujiTrabuco 5 GTX           Asics        83.0   
824  Reebok All Terrain Thunder 2.0          Reebok        82.0   
825              Inov-8 Mudclaw 265          Inov-8        82.0   
826                   Zoot Ovwa 2.0            Zoot        82.0   
827      Salomon S-Lab Spikecross 3         Salomon        82.0   
828        Under Armour Charge RC 2    Under Armour        82.0   
829        Adidas Adizero Feather 4          Adidas        70.0   
830    Under Armour Micro G Nitrous    Under Armour        81.0   
831       Merrell All Out Terra Ice         Merrell        81.0   
832             Salomon S-Lab Sonic         Salomon        81.0   
833             Asics FuzeX Rush CM           Asics        81.0   
834              Puma Faas 100 R v3            Puma        80.0   
835         Asics Gel Kenun Shinkai           Asics        80.0   
836     The North Face Ultra TR III  The North Face        80.0   
837                 Inov-8 Oroc 340          Inov-8        79.0   
838              Zoot Advantage 3.0            Zoot        79.0   
839         Inov-8 Road-X-Treme 220          Inov-8        79.0   
840         Adidas Pure Boost Chill          Adidas        79.0   
841                      Zoot Diego            Zoot        78.0   
842              Saucony Ride 9 GTX         Saucony        78.0   
843              Skechers GOtrail 2        Skechers        77.0   
844             Adidas Adizero XT 5          Adidas        76.0   
845           Mizuno Wave Cruise 10          Mizuno        64.0   
846        

In [85]:
# Binarize features.
df['discontinued'] = df['discontinued'].map({'Y': 1, 'N': 0})
df['terrain'] = df['terrain'].map({'Road': 1, 'Trail': 0})
df['use'] = df['use'].map({'Daily running': 1, 'Competition': 0})
df['waterproof'] = df['waterproof'].map({'Waterproof': 1, 'Not waterproof': 0})
df['water_repellent'] = df['water_repellent'].map({'Water repellent': 1, 'Not water repellent': 0})
df['maximalist'] = df['maximalist'].map({'Maximalist': 1, 'Not maximalist': 0})
df['minimalist'] = df['minimalist'].map({'Minimalist': 1, 'Not minimalist': 0})
df['triathlon'] = df['triathlon'].map({'Triathlon': 1, 'Not triathlon': 0})

In [86]:
df

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike        94.0   
1                   Brooks Ghost 11          Brooks        94.0   
2               Asics Gel Kayano 25           Asics        95.0   
3               Asics Gel Venture 6           Asics        94.0   
4              Hoka One One Bondi 6    Hoka One One        91.0   
5             Merrell Trail Glove 4         Merrell        93.0   
6          Nike Air Zoom Pegasus 36            Nike        91.0   
7          Brooks Adrenaline GTS 18          Brooks        94.0   
8           Nike Zoom Pegasus Turbo            Nike        92.0   
9               Asics Gel Nimbus 21           Asics        91.0   
10               Brooks Glycerin 16          Brooks        93.0   
11               Brooks Glycerin 17          Brooks        94.0   
12           Hoka One One Clifton 5    Hoka One One        93.0   
13        Nike Free RN Flyknit 2018            Nike        94.0   
14        Nike Epic React Flyknit 2            Nike        89.0   
15             Adidas Pure Boost Go          Adidas        94.0   
16             Adidas Ultraboost 19          Adidas        95.0   
17               Adidas Solar Boost          Adidas        94.0   
18            Nike Zoom Fly Flyknit            Nike        92.0   
19              Altra Escalante 1.5           Altra        94.0   
20         Hoka One One Speedgoat 3    Hoka One One        97.0   
21         Nike Air Zoom Pegasus 34            Nike        92.0   
22                Hoka One One Cavu    Hoka One One        92.0   
23                Nike Free RN 2018            Nike        92.0   
24             Mizuno Wave Rider 22          Mizuno        94.0   
25                     Brooks Revel          Brooks        95.0   
26   New Balance Fresh Foam 1080 v9     New Balance        95.0   
27                  Brooks Ghost 10          Brooks        95.0   
28               Saucony Kinvara 10         Saucony        93.0   
29             Asics Gel Quantum 90           Asics        92.0   
..                              ...             ...         ...   
818       Reebok All Terrain Thrill          Reebok        72.0   
819              Altra Paradigm 1.5           Altra        83.0   
820              Altra Instinct 3.5           Altra        83.0   
821        Brooks Adrenaline ASR 12          Brooks        83.0   
822         Asics Gel FujiTrainer 3           Asics        83.0   
823     Asics Gel FujiTrabuco 5 GTX           Asics        83.0   
824  Reebok All Terrain Thunder 2.0          Reebok        82.0   
825              Inov-8 Mudclaw 265          Inov-8        82.0   
826                   Zoot Ovwa 2.0            Zoot        82.0   
827      Salomon S-Lab Spikecross 3         Salomon        82.0   
828        Under Armour Charge RC 2    Under Armour        82.0   
829        Adidas Adizero Feather 4          Adidas        70.0   
830    Under Armour Micro G Nitrous    Under Armour        81.0   
831       Merrell All Out Terra Ice         Merrell        81.0   
832             Salomon S-Lab Sonic         Salomon        81.0   
833             Asics FuzeX Rush CM           Asics        81.0   
834              Puma Faas 100 R v3            Puma        80.0   
835         Asics Gel Kenun Shinkai           Asics        80.0   
836     The North Face Ultra TR III  The North Face        80.0   
837                 Inov-8 Oroc 340          Inov-8        79.0   
838              Zoot Advantage 3.0            Zoot        79.0   
839         Inov-8 Road-X-Treme 220          Inov-8        79.0   
840         Adidas Pure Boost Chill          Adidas        79.0   
841                      Zoot Diego            Zoot        78.0   
842              Saucony Ride 9 GTX         Saucony        78.0   
843              Skechers GOtrail 2        Skechers        77.0   
844             Adidas Adizero XT 5          Adidas        76.0   
845           Mizuno Wave Cruise 10          Mizuno        64.0   
846        

In [ ]:
# Try recalculating user_rating to get a more exact float
# df[['user_5star', 'user_4star', 'user_3star', 'user_2star', 'user_1star']].mean(axis=1)

# Observed that the percentages do not add up exactly to 100% due to rounding errors (e.g. index 23)
# Hence, this recalculation of user_rating will not be done.

In [99]:
# competition_mask = df['use'] == 0 # Competition running use
# df[competition_mask].sort_values(by='weight_m', ascending=False)

# daily_mask = df['use'] == 1 # Daily running use
# df[daily_mask].sort_values(by='weight_m', ascending=True)

# df.sort_values(by='toe_drop_m', ascending=True)

In [93]:
# Check column types.
df.dtypes

name                    object
brand                   object
core_score             float64
user_review_count      float64
expert_review_count    float64
user_rating            float64
expert_rating          float64
user_5star              object
user_4star              object
user_3star              object
user_2star              object
user_1star              object
discontinued             int64
terrain                  int64
arch_support            object
use                      int64
waterproof               int64
water_repellent          int64
maximalist               int64
minimalist               int64
triathlon                int64
price                  float64
weight                  object
toe_drop                object
heel_height             object
forefoot_height         object
width                   object
good_summary            object
bad_summary             object
summary                 object
info                    object
size_fit                object
outsole 

In [94]:
# Change column types.
df = df.astype({
    'core_score': int,
    'user_review_count': int,
    'expert_review_count': int,
    'user_5star': int,
    'user_4star': int,
    'user_3star': int,
    'user_2star': int,
    'user_1star': int,
    'price': int,
    'weight_m': int,
    'weight_w': int,
    'toe_drop_m': int,
    'toe_drop_w': int,
    'heel_height_m': int,
    'heel_height_w': int,
    'forefoot_height_m': int,
    'forefoot_height_w': int
})

In [95]:
# Create categorical features.
df['zero_drop_m'] = df['toe_drop_m'].map(lambda x: 1 if x==0 else 0)
df['zero_drop_w'] = df['toe_drop_w'].map(lambda x: 1 if x==0 else 0)
df['low_drop_m'] = df['toe_drop_m'].map(lambda x: 1 if x>=1 and x<=4 else 0)
df['low_drop_w'] = df['toe_drop_w'].map(lambda x: 1 if x>=1 and x<=4 else 0)

In [96]:
# Save cleaned data to csv.
df.to_csv('shoe_data_cleaned.csv', index=False)

In [97]:
# Save to pickle file.
pickle.dump(df, open('shoe_data.pkl', 'wb'))

In [98]:
# Check ...
pickle.load(open('shoe_data.pkl', 'rb'))

name           brand  core_score  \
0          Nike Air Zoom Pegasus 35            Nike          94   
1                   Brooks Ghost 11          Brooks          94   
2               Asics Gel Kayano 25           Asics          95   
3               Asics Gel Venture 6           Asics          94   
4              Hoka One One Bondi 6    Hoka One One          91   
5             Merrell Trail Glove 4         Merrell          93   
6          Nike Air Zoom Pegasus 36            Nike          91   
7          Brooks Adrenaline GTS 18          Brooks          94   
8           Nike Zoom Pegasus Turbo            Nike          92   
9               Asics Gel Nimbus 21           Asics          91   
10               Brooks Glycerin 16          Brooks          93   
11               Brooks Glycerin 17          Brooks          94   
12           Hoka One One Clifton 5    Hoka One One          93   
13        Nike Free RN Flyknit 2018            Nike          94   
14        Nike Epic React Flyknit 2            Nike          89   
15             Adidas Pure Boost Go          Adidas          94   
16             Adidas Ultraboost 19          Adidas          95   
17               Adidas Solar Boost          Adidas          94   
18            Nike Zoom Fly Flyknit            Nike          92   
19              Altra Escalante 1.5           Altra          94   
20         Hoka One One Speedgoat 3    Hoka One One          97   
21         Nike Air Zoom Pegasus 34            Nike          92   
22                Hoka One One Cavu    Hoka One One          92   
23                Nike Free RN 2018            Nike          92   
24             Mizuno Wave Rider 22          Mizuno          94   
25                     Brooks Revel          Brooks          95   
26   New Balance Fresh Foam 1080 v9     New Balance          95   
27                  Brooks Ghost 10          Brooks          95   
28               Saucony Kinvara 10         Saucony          93   
29             Asics Gel Quantum 90           Asics          92   
..                              ...             ...         ...   
818       Reebok All Terrain Thrill          Reebok          72   
819              Altra Paradigm 1.5           Altra          83   
820              Altra Instinct 3.5           Altra          83   
821        Brooks Adrenaline ASR 12          Brooks          83   
822         Asics Gel FujiTrainer 3           Asics          83   
823     Asics Gel FujiTrabuco 5 GTX           Asics          83   
824  Reebok All Terrain Thunder 2.0          Reebok          82   
825              Inov-8 Mudclaw 265          Inov-8          82   
826                   Zoot Ovwa 2.0            Zoot          82   
827      Salomon S-Lab Spikecross 3         Salomon          82   
828        Under Armour Charge RC 2    Under Armour          82   
829        Adidas Adizero Feather 4          Adidas          70   
830    Under Armour Micro G Nitrous    Under Armour          81   
831       Merrell All Out Terra Ice         Merrell          81   
832             Salomon S-Lab Sonic         Salomon          81   
833             Asics FuzeX Rush CM           Asics          81   
834              Puma Faas 100 R v3            Puma          80   
835         Asics Gel Kenun Shinkai           Asics          80   
836     The North Face Ultra TR III  The North Face          80   
837                 Inov-8 Oroc 340          Inov-8          79   
838              Zoot Advantage 3.0            Zoot          79   
839         Inov-8 Road-X-Treme 220          Inov-8          79   
840         Adidas Pure Boost Chill          Adidas          79   
841                      Zoot Diego            Zoot          78   
842              Saucony Ride 9 GTX         Saucony          78   
843              Skechers GOtrail 2        Skechers          77   
844             Adidas Adizero XT 5          Adidas          76   
845           Mizuno Wave Cruise 10          Mizuno          64   
846        